Makefile

```diff
@@ -0,0 +1,55 @@
+include .env
+BINARY := oauth2_proxy
+VERSION := $(shell git describe --always --long --dirty --tags 2>/dev/null || echo "undefined")
+
+.PHONY: all
+all: dep lint $(BINARY)
+
+.PHONY: clean
+clean:
+	rm -rf release
+	rm -f $(BINARY)
+
+.PHONY: distclean
+distclean: clean
+	rm -rf vendor
+
+BIN_DIR := $(GOPATH)/bin
+GOMETALINTER := $(BIN_DIR)/gometalinter
+
+$(GOMETALINTER):
+	$(GO) get -u github.com/alecthomas/gometalinter
+	gometalinter --install %> /dev/null
+
+.PHONY: lint
+lint: $(GOMETALINTER)
+	$(GOMETALINTER) --vendor --disable-all \
+		--enable=vet \
+		--enable=vetshadow \
+		--enable=golint \
+		--enable=ineffassign \
+		--enable=goconst \
+		--enable=deadcode \
+		--enable=gofmt \
+		--enable=goimports \
+		--tests ./...
+
+.PHONY: dep
+dep:
+	$(DEP) ensure --vendor-only
+
+.PHONY: build
+build: clean $(BINARY)
```

Comment:
If the user has `MAKEFLAGS=-j4` or something like that in their environment (to enable parallel builds using multiple CPU cores) then the prerequisites "clean" and "$(BINARY)" will be run in parallel, not in order.

(This also affects `release: deb lint test` below.)

Reply:
> If .NOTPARALLEL is mentioned as a target, then this invocation of make will be run serially, even if the ‘-j’ option is given. Any recursively invoked make command will still run recipes in parallel (unless its makefile also contains this target). Any prerequisites on this target are ignored. 

That will work, and apply to the whole Makefile not just the "build" target, which is fine for this kind of Makefile (but it's a bit confusing that you list "build" there).

Reply:
I missed the fact that this is global 🤦‍♂️ I'm not a make expert! I have moved the `.NOTPARALLEL` to the top and left it at that, unless you have a decent suggestion for making sure `build` doesn't get run in parallel, everything else should be fine in parallel

Reply:
`.NOTPARALLEL` is not really a core "make" concept, I was not familiar with it myself before ... but `make` is an interest of mine :)

When most targets are `PHONY`, things get awkward, because make was designed for targets which are files, which depend on other files. So "clean" is particularly awkward, because the goal is the *absense* of files, and it conflicts with other targets.

In the `Dockerfile` you do:

```
RUN ./configure && make clean oauth2_proxy
```

and that too is problematic, that says to try to satisfy "clean" and "oauth2_proxy" targets, at the same time (but without parallelism it'll be one then the other, but it could be confused if there's a common dependency which it assumes it already did). It's conceptually better to do

```
RUN ./configure && make clean && make oauth2_proxy
```

or make a `clean-build` target like so (you tried something a bit like this before):

```
clean-build:
	$(MAKE) clean
	$(MAKE) $(BINARY)
```

So those are sequential steps instead of parallel dependencies. Basically, if you're using `make` as just a scriptlets-runner, don't use prerequisites/dependencies.

One reason you always seem to need to "clean" before building the binary, is that if the binary exists then make assumes it is up-to-date, because it has no prerequisites. They could be added:

```
$(BINARY): $(wildcard *.go api/*.go cookie/*.go providers/*.go)
	$(GO) build ...
```

That wouldn't include the vendored libraries. You could add the `dep` step to the prerequisites list, but then it would always build, because `dep` is a phony target ... so don't bother.

All that to say, `.NOTPARALLEL` at the top is as good a way as any to satisfy this overall problem :)